modules/core/src/main/java/org/apache/ignite/internal/client/thin/ClientCacheEntryListenerHandler.java

```diff
@@ -0,0 +1,266 @@
+            BinaryOutputStream out = payloadCh.out();
+
+            out.writeInt(ClientUtils.cacheId(jCacheAdapter.getName()));
+            out.writeByte(keepBinary ? KEEP_BINARY_FLAG_MASK : 0);
+            out.writeInt(pageSize);
+            out.writeLong(timeInterval);
+            out.writeBoolean(includeExpired);
+
+            if (rmtFilterFactory == null)
+                out.writeByte(GridBinaryMarshaller.NULL);
+            else {
+                utils.writeObject(out, rmtFilterFactory);
+                out.writeByte((byte)1); // Java platform
+            }
+        };
+
+        ch.addChannelCloseListener(chCloseLsnr);
+
+        try {
+            T2<ClientChannel, Long> params = ch.service(
+                ClientOperation.QUERY_CONTINUOUS,
+                qryWriter,
+                res -> new T2<>(res.clientChannel(), res.in().readLong())
+            );
+
+            clientCh = params.get1();
+            rsrcId = params.get2();
+        }
+        catch (ClientError e) {
+            ch.removeChannelCloseListener(chCloseLsnr);
+
+            throw new ClientException(e);
+        }
+
+        clientCh.addNotificationListener(this);
+    }
+
+    /**
+     * @param ch Channel.
+     */
+    private void onChannelClosed(ClientChannel ch) {
+        if (ch == clientCh) {
+            if (locLsnr instanceof ClientChannelDisconnectListener)
+                ((ClientChannelDisconnectListener)locLsnr).onDisconnected();
+
+            U.closeQuiet(this);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void acceptNotification(
+            ClientChannel ch,
+            ClientOperation op,
+            long rsrcId,
+            ByteBuffer payload,
+            Exception err
+    ) {
+        if (op == ClientOperation.QUERY_CONTINUOUS_EVENT_NOTIFICATION && rsrcId == this.rsrcId) {
+            if (err == null && payload != null) {
+                BinaryInputStream in = BinaryByteBufferInputStream.create(payload);
+
+                int cnt = in.readInt();
+
+                List<CacheEntryEvent<? extends K, ? extends V>> evts = new ArrayList<>(cnt);
+
+                for (int i = 0; i < cnt; i++) {
+                    K key = utils.readObject(in, keepBinary);
+                    V oldVal = utils.readObject(in, keepBinary);
+                    V val = utils.readObject(in, keepBinary);
+                    byte evtTypeByte = in.readByte();
+
+                    EventType evtType = eventType(evtTypeByte);
+
+                    if (evtType != null) // Skip unknown event types.
+                        evts.add(new CacheEntryEventImpl<>(jCacheAdapter, evtType, key, oldVal, val));
+                }
+
+                locLsnr.onUpdated(evts);
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public synchronized void close() {
+        ch.removeChannelCloseListener(chCloseLsnr);
+
+        if (clientCh != null) {
+            clientCh.removeNotificationListener(this);
```

Comment:
Even though we never set `clientCh` from non-null to null, and this code seems to be safe right now,
the field is still marked `volatile`, so this use-after-nullcheck looks suspicious.

To reduce cognitive load, copy `clientCh` to a temp variable, and then check for null.